storage/storage-interface/src/finality_view.rs

```diff
@@ -0,0 +1,159 @@
+use std::sync::RwLock;
+
+use aptos_crypto::HashValue;
+use aptos_types::{
+    aggregate_signature,
+    block_info::BlockInfo,
+    ledger_info::{LedgerInfo, LedgerInfoWithSignatures},
+    transaction::Version,
+};
+
+use crate::{AptosDbError, DbReader, Result};
+
+/// A wrapper over [`DbReader`], representing ledger at the end of a block
+/// as the latest ledger and its version as the latest transaction version.
+pub struct FinalityView<Db> {
+    reader: Db,
+    finalized_ledger_info: RwLock<Option<LedgerInfoWithSignatures>>,
+}
+
+impl<Db> FinalityView<Db> {
+    pub fn new(reader: Db) -> Self {
+        Self {
+            reader,
+            finalized_ledger_info: RwLock::new(None),
+        }
+    }
+}
+
+impl<Db: DbReader> FinalityView<Db> {
+    /// Updates the information on the latest finalized block at the specified height.
+    pub fn set_finalized_block_height(&self, height: u64) -> Result<()> {
+        let (_start_ver, end_ver, block_event) = self.reader.get_block_info_by_height(height)?;
+        let block_hash = block_event.hash()?;
+        let block_info = BlockInfo::new(
+            block_event.epoch(),
+            block_event.round(),
+            block_hash,
+            self.reader.get_accumulator_root_hash(end_ver)?,
+            end_ver,
+            block_event.proposed_time(),
+            None,
+        );
+        // Sanity checks: finalization should not be set on an empty database,
+        // the finality version should not exceed the latest committed.
+        match self.reader.get_latest_state_checkpoint_version()? {
+            None => return Err(AptosDbError::Other("no ledger states to finalize".into())),
```

Comment:
So, we _are_ relying on checkpoints?

Reply:
All blocks in _opt_ should be committed to storage by the opt executor, so it's the best sanity check I could come up with that does not require explicit sync-ups from the opt state.
In another change @0xmovses added the code above to fetch block info; I'll check if it already ensures that we get it from a committed block and this check is redundant.